Smart Home: Camera viewing and input streaming

garbear opened the pull request:
## Description

This PR introduces a new paradigm of Kodi usage: smart home interaction and control. Camera viewing and input streaming are provided as two initial use cases.

The approach I've taken with smart home is a bit unique: I built on ROS 2. While this introduces some heavy overhead, building on an industrial decentralized communication framework allows for scalability to virtually unlimited low power smart home devices. My smart home's computation graph is currently at 8 Linux nodes and 3 Arduino nodes and growing.

Alongside this PR is a repo devoted to my smart home OS: https://github.com/eigendude/OASIS.

OASIS provides a complete ROS 2 stack for computer vision, input streaming and general automation. It also provides a complete implementation of the Firmata protocol for communicating with Arduinos, with additional support for temperature and humidity sensors, I2C, servos, sonar, SPI, stepper motors, and 4-wire CPU fans.

I'm not seriously proposing we merge this and add ROS 2 as a dependency; it adds 2 millions lines of code. I'm just sharing the code I run everyday at home, and maybe it'll inspire someone.

### Camera viewing

A new `cameraview` control is introduced. It takes a parameter, the ROS image topic to subscribe to. For testing, I created the following `SmartHome.xml` window:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<window>
	<controls>
		<control type="cameraview">
			<width>50%</width>
			<height>50%</height>
			<topic>/oasis/netbook/image_raw</topic>
		</control>
		<control type="cameraview">
			<left>50%</left>
			<height>50%</height>
			<topic>/oasis/lenovo/image_raw</topic>
		</control>
		<control type="cameraview">
			<width>50%</width>
			<top>50%</top>
			<topic>/oasis/netbook/foreground</topic>
		</control>
		<control type="cameraview">
			<left>50%</left>
			<top>50%</top>
			<topic>/oasis/lenovo/foreground</topic>
		</control>
	</controls>
</window>
```

This window creates a 2x2 matrix. The top two cameraview controls are the direct output of two laptop cameras. The bottom two cameraview controls are the output of a rudimentary person detector using a [background subtraction library](https://github.com/andrewssobral/bgslibrary) applied to the camera feeds.

Screenshot of the above window (not bad for low light netbook cameras):

![Not very good person tracker](https://user-images.githubusercontent.com/531482/159408776-23aef652-56cc-47bf-97dd-a4fe497852c8.png)

I forked the [Kinect 2 driver](https://github.com/eigendude/oasis_kinect2) and ported it to ROS 2. You can see depth registration working, but I'll probably do skeletal tracking on the 2D image via MediaPipe as a person detection solution.

![Smart Home Camera View](https://user-images.githubusercontent.com/531482/159824434-7f053f01-e43b-4934-b105-b006c72df4fd.png)

### Input streaming

The second use case is input streaming for smart home control. Kodi advertises two ROS topics:

* `/peripherals` - A list of attached input peripherals, which is all controllers plus the keyboard and mouse
* `/input` - A topic that publishes the state of controllers

Kodi also provides a ROS service:

* `/capture_input` - Calling this service causes a peripheral's input to be captured and sent over the `input` topic.

In my home, I'm using a PS controller to drive my childhood Lego train. The input is captured by Kodi running on my NAS with a Bluetooth dongle, and sent to a Raspberry Pi with an Ardiuno and a robotics motor controller that drives the train's 9V motors.

[![Lego Train via PS4 Controller](https://img.youtube.com/vi/zMA9HYPH4Tw/hqdefault.jpg)](https://youtu.be/zMA9HYPH4Tw)

## Motivation and context

Smart Home!!!!

## How has this been tested?

I've been using Kodi as part of my smart home setup since May of 2021. ROS 2 is so robust that my system barely goes down.

## What is the effect on users?

* Smart Home!!!

## Related PRs

Depends on:

* https://github.com/xbmc/xbmc/pull/21182
* https://github.com/a9183756-gh/Arduino-CMake-Toolchain/pull/55
* https://github.com/paul-shuvo/iai_kinect2_opencv4/pull/3
* https://github.com/MrYsLab/pymata-express/pull/40
* https://github.com/xbmc/peripheral.joystick/pull/235
* https://github.com/eigendude/OASIS

## Types of change

<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

garbear committed:
[temp] CMake: Also search for unmodified casing

garbear committed:
Smart Home: Camera and input streaming

garbear committed:
[temp] Remove uninstall target

Error was:

  CMake Error at CMakeLists.txt:479 (add_custom_target):
    add_custom_target cannot create target "uninstall" because another target
    with the same name already exists.  The existing target is a custom target
    created in source directory.  See documentation for policy CMP0002 for
    more details.

basilgello commented:
Wow!! Definitely interesting research!

garbear committed:
Smart Home: LEGO train power station HUD

This commit introduces a ROS 2 subscriber to the train power station
state, and exposes the infomation to the skin.

The HUD is implemented in the service.oasis add-on.

garbear committed:
Estuary: Add 120pt font

garbear committed:
[temp] Increase refresh rate of list provider items

garbear committed:
[temp] Disable fullscreen on movie start

basilgello commented:
> I'm not seriously proposing we merge this and add ROS 2 as a dependency; it adds 2 millions lines of code.

Cannot we make only required changes to core and develop everything else in an addon depending on ROS?

garbear commented:
> Cannot we make only required changes to core and develop everything else in an addon depending on ROS?

It's possible we can move the ROS dependency to a binary add-on. But it would easily triple the amount of code. Unfortunately I'm just not at a stage in life where I can pull another RetroPlayer and upstream 40K lines of code :wink:

Maybe I could make ROS 2 an optional dependency, but that still leaves us with the fact that ROS 2 is hard to install, much more so when you have to compile from source. IMO I would never recommend a smart home enthusiast to go through this process. You pretty much have to be a dedicated hardcore roboticist to power through and use ROS 2. As a result, this is really more a "domitics" (domestic robotics) project and less a "smart home" project.

But I'll keep working on it, as technology can change and become more accessible and maybe the day will finally arrive for this to be included in Kodi.

garbear commented:
Rebased after merging https://github.com/xbmc/xbmc/pull/21182

garbear committed:
Smart Home: LEGO train HUD

This commit introduces a ROS 2 subscriber to the LEGO train state,
and exposes information to the skin.

garbear committed:
Smart Home: LEGO air quality lab HUD

garbear commented:
I updated the code after making ROS optional and rebasing on the following PRs:

* https://github.com/xbmc/xbmc/pull/22349
* https://github.com/xbmc/xbmc/pull/22633
* https://github.com/xbmc/xbmc/pull/22861

Now that ROS is optional, Jenkins should be able to handle this. Jenkins build this please.

garbear commented:
Jenkins build this please

garbear commented:
jenkins build this please

garbear committed:
Add service.oasis Python service

garbear commented:
jenkins build this please

basilgello commented:
@garbear you keep pushing smarthome branches to xbmc/xbmc directly. @fuzzard cleaned them but there is something with your `git remote` setup. Please fix :)

garbear committed:
[temp] Disable service.xbmc.versioncheck

garbear committed:
[temp] Disable notifications

garbear commented:
Not stale, still actively developed, just haven't pushed Alpha 3 based branch yet.
